presto-main/src/main/java/com/facebook/presto/operator/SpatialJoinOperator.java

```diff
@@ -0,0 +1,272 @@
+    private final PageBuilder pageBuilder;
+    @Nullable
+    private Page probe;
+
+    // The following fields represent the state of the operator in case when processProbe yielded or
+    // filled up pageBuilder before processing all records in a probe page.
+    private int probePosition;
+    @Nullable
+    private long[] joinAddresses;
+    private int nextJoinAddress;
+
+    private boolean finishing;
+    private boolean finished;
+
+    public SpatialJoinOperator(
+            OperatorContext operatorContext,
+            List<Type> outputTypes,
+            List<Type> probeTypes,
+            List<Integer> probeOutputChannels,
+            int probeGeometryChannel,
+            PagesSpatialIndexFactory pagesSpatialIndexFactory)
+    {
+        this.operatorContext = operatorContext;
+        this.probeTypes = ImmutableList.copyOf(probeTypes);
+        this.probeOutputChannels = ImmutableList.copyOf(probeOutputChannels);
+        this.probeGeometryChannel = probeGeometryChannel;
+        this.pagesSpatialIndexFactory = pagesSpatialIndexFactory;
+        this.pagesSpatialIndexFuture = pagesSpatialIndexFactory.createPagesSpatialIndex();
+        this.outputTypes = ImmutableList.copyOf(outputTypes);
+        this.pageBuilder = new PageBuilder(outputTypes);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return outputTypes;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished && pagesSpatialIndexFuture.isDone() && !pageBuilder.isFull() && probe == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        verify(probe == null);
+        probe = page;
+        probePosition = 0;
+
+        joinAddresses = null;
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        verify(!finished);
+        if (!pageBuilder.isFull() && probe != null) {
+            processProbe();
+        }
+
+        if (pageBuilder.isFull()) {
+            Page page = pageBuilder.build();
+            pageBuilder.reset();
+            return page;
+        }
+
+        if (finishing && probe == null) {
+            Page page = null;
+            if (!pageBuilder.isEmpty()) {
+                page = pageBuilder.build();
+                pageBuilder.reset();
+            }
+            pagesSpatialIndexFactory.probeOperatorFinished();
+            pagesSpatialIndexFuture = null;
+            finished = true;
+            return page;
+        }
+
+        return null;
+    }
+
+    private void processProbe()
+    {
+        verify(probe != null);
+
+        PagesSpatialIndex pagesSpatialIndex = getDone(pagesSpatialIndexFuture);
+        DriverYieldSignal yieldSignal = operatorContext.getDriverContext().getYieldSignal();
+        for (; probePosition < probe.getPositionCount(); probePosition++) {
```

Comment:
Consider  `while (probePosition < probe.getPositionCount())` + putting `probePosition++` at the loop end, where you already `this.joinAddresses = null;`